GGRC-4429 Refactor log event procedure for nightly cron job

mashabarashko opened the pull request:
<!-- If your PR depends on other tickets or PRs, you can list them here
# Dependencies

This PR is `on hold` until the dependencies are merged:

- [ ] GGRC-1234
- [ ] #1234
-->

# Issue description

Recently start_recurring_cycles function of the nightly cron jobs has been refactored so that many commits were performed instead of a single one, and multiple events were logged (each one before a commit). However, only one event has to be logged in this case.

# Steps to test the changes

1. Run start_recurring_cycles() function on the qa db dump;
2. Check the number of created events in the db (1 event should be created);
3. Check the number of revisions created (should be non-zero number).

# Solution description

One more argument - event object - has been added to the log_event function.The next situations are possible:

1. Revisions list is empty and event is None. In this case nether event nor revisions won't be created.
2. Revisions list is not empty and event is None. In this case event and appropriate revisions will be created.
3. Revisions list is not empty and event is not None. In this case appropriate revisions will be created, while event won't be changed.
4. Revisions list is empty and event is not None. In this case event will not be changed, and new revisions will not be added.

The first two cases reflect the previously used logic, the last two are handled during start_recurring_cycles() execution.

Only one event is created in the start_recurring_cycles() now, and new revisions are attached to this event during looping through workflows.

# Sanity checklist

- [x] I have clicked through the app to make sure my changes work and not break the app.
- [x] I have applied the correct milestone and labels.
- [x] My changes fix the issue described in the description (and do nothing else). 🤞
- [x] My changes are covered by tests.
- [ ] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [x] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [x] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

<!-- If your PR includes a migration include the additional checklist items
# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)
-->

# PR Review checklist

- [ ] The changes fix the issue and don't cause any apparent regressions.
- [ ] Labels and milestone are correctly set.
- [ ] The solution description matches the changes in the code.
- [ ] There is no apparent way to improve the performance & design of the new code.
- [ ] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

# TODO

- [x] Add revisions generation

ahetmanski commented:
`needs work` cause PR description TODO section.

ahetmanski commented:
@mashabarashko fix codestyle please.

pod2metra commented on src/ggrc/converters/base_block.py:
```diff
@@ -608,7 +608,7 @@ def save_import(self):
     try:
       modified_objects = get_modified_objects(db.session)
-      import_event = log_event(db.session, None)
+      import_event = log_event(db.session, obj=None)
```
You may change this one to log_event(db.session)

pod2metra commented on src/ggrc/utils/log_event.py:
```diff
@@ -61,24 +61,27 @@ def _get_log_revisions(current_user_id, obj=None, force_obj=False):
 
-def log_event(session, obj=None, current_user_id=None, flush=True,
-              force_obj=False):
+def log_event(session, **kwargs):
```
why you didn't use 
log_event(session, obj=None, current_user_id=None, flush=True, force_obj=False, event=None)?

pod2metra commented on src/ggrc/utils/user_generator.py:
```diff
@@ -54,7 +58,7 @@ def create_user(email, **kwargs):
   db.session.add(user)
   db.session.flush()
-  log_event(db.session, user, user.id)
+  log_event(db.session, obj=user, current_user_id=user.id)
```
here is the double flush. 
on row 60 and in log_event procedure.

pod2metra commented on test/integration/ggrc_workflows/test_recurring_cycles_event.py:
```diff
@@ -0,0 +1,68 @@
+
+    self.assertEqual(wf_models.Cycle.query.count() - self.cycles_num, 2)
+    self.assertEqual(models.Event.query.count() - self.event_num, 1)
+    self.assertTrue(models.Revision.query.count() - self.revision > 0)
```
please use assertEqual(num_of_created_revisions, models.Revision.query.count()) 
Here you check only that revision created even 1 revision. But you should check that revision created to each required object and without duplicates.  
Also you may check number of revisions that was related to newly cerated Event.

mashabarashko commented:
@pod2metra I fixed integration tests so that now exact revisions number for particular event is checked. I also reverted commit regarding log_event arguments list refactoring and put pylint disable instead.

mashabarashko commented:
@pod2metra fixed adding event to session in the loop.

zidarsk8 commented on test/integration/ggrc_workflows/test_recurring_cycles_event.py:
```diff
@@ -0,0 +1,71 @@
+
+    events = models.Event.query.filter(models.Event.action == 'BULK').all()
+
+    self.assertEqual(wf_models.Cycle.query.count() - self.cycles_num, 2)
```
@mashabarashko These Assertions are really hard to read,

From logic perspective we are asserting that the new cycle count has 2 more cycles than the old count so it would be more readable if this were written as
```python
self.assertEqual(wf_models.Cycle.query.count(), self.cycles_num + 2)
```

mashabarashko replied:
@zidarsk8 fixed.

zidarsk8 commented on test/integration/ggrc_workflows/test_recurring_cycles_event.py:
```diff
@@ -0,0 +1,71 @@
+
+    self.assertEqual(wf_models.Cycle.query.count() - self.cycles_num, 0)
+    self.assertEqual(len(events), 0)
+    self.assertEqual(models.Revision.query.count() - revisions_count, 0)
```
And this assertion is too cryptic for checking if count A equals count B.

mashabarashko replied:
@zidarsk8 I agree, fixed.
